industrial_taxonomy/pipeline/glass_house/flow.py

```diff
@@ -0,0 +1,196 @@
+        full_top_matches: Full match results
+    """
+
+    test_mode = Parameter(
+        "test-mode",
+        help="Whether to run in test mode (on a small subset of data)",
+        type=bool,
+        default=lambda _: not current.is_production,
+    )
+    should_clean_names = Parameter(
+        "clean-names",
+        help="Whether to clean names using Jacchammer default `preproc_names`",
+        type=bool,
+        default=False,
+    )
+
+    names_x: "DataFrame"
+    names_y: "DataFrame"
+    clean_names_x: "DataFrame"
+    clean_names_y: "DataFrame"
+    run_id: int
+    full_top_matches: List[FullMatchResult]  # TODO: py37 error
+
+    @step
+    def start(self):
+        """Load raw data"""
+        from industrial_taxonomy.getters.companies_house import (
+            get_name as get_companies_house_names,
+        )
+        from industrial_taxonomy.getters.glass import (
+            get_organisation as get_glass_organisation,
+        )
+
+        self.run_id = current.origin_run_id or current.run_id
+
+        nrows = 10_000 if (self.test_mode and not current.is_production) else None
+
+        self.names_x = get_companies_house_names().name.head(nrows)
+        self.names_y = get_glass_organisation().name.head(nrows)
+
+        self.next(self.clean_names)
+
+    @pip(path="requirements.txt")
+    @step
+    def clean_names(self):
+        """Pre-process names if `self.should_clean_names`."""
+        from jacc_hammer.name_clean import preproc_names
+
+        self.clean_names_x = (
+            preproc_names(self.names_x) if self.should_clean_names else self.names_x
+        )
+        self.clean_names_y = (
+            preproc_names(self.names_y) if self.should_clean_names else self.names_y
+        )
+
+        self.next(self.match)
+
+    @resources(memory=64_000)
+    @pip(path="requirements.txt")
+    @step
+    def match(self):
+        """The core fuzzy matching algorithm"""
+        from itertools import starmap as starmap_
+        from pathlib import Path
+        from tempfile import TemporaryDirectory
+
+        from jacc_hammer.fuzzy_hash import (
+            Cos_config,
+            Fuzzy_config,
+            match_names_stream,
+        )
+        from pandas import DataFrame
+        from toolz.curried import curry, pipe
+        from toolz.curried import map as tmap
+
+        from utils import optimise_chunk
+
+        starmap = curry(starmap_)
+
+        def persist_chunk(chunk_id: int, chunk: DataFrame) -> int:
+            key = f"match-chunks-{chunk_id}"
+            with S3(run=self) as s3:
+                s3.put(key, chunk.to_csv())
+            print(f"Persisted chunk {chunk_id} of shape {chunk.shape}")
+            return chunk_id
+
+        tmp_dir = Path(TemporaryDirectory(dir=".").name)
+        tmp_dir.mkdir()
+
+        cos_config = Cos_config()
+        fuzzy_config = Fuzzy_config(num_perm=128)
+        match_config = dict(
+            threshold=50,
+            chunksize=100_000,
```

Comment:
I'm a bit unclear on the rationale for when constants/parameters should go in the body of a script, be set as global or go in a config (I think the latter is for flor params)?

Reply:
Me too :stuck_out_tongue: 
I think it's generally quite a subjective decision apart from a few clear-cut cases

Also, don't assume because I wrote the code up there  :point_up:  that it's the right way


2 questions that if the answer to either is "Yes" suggest it should be in a config file:
1) Does this variable only relate to this pipeline (i.e. is it shared across other utils/flows/getters)
2) Does the variable relate closely to the problem domain
    - "Is this a good enough match between two companies?" - close to the problem domain
    - "Should I have 128 or 256 trees in my ensemble?" - not close to the problem domain

### Based on the above:

I think `threshold` _**could**_ be a param...
It can map to the match threshold in the getter but on the other hand it exists here as a memory optimisation that we make before choosing the final match threshold to use for the rest of the analysis.
What should be done is that this should be stored as a data artifact so that when a getter chooses the match threshold we can check that `getter threshold` (the domain decision)  `>=` `flow threshold` (optimisation decision)

The others are "fine" - perhaps they are better as constants at the top of the file with a comment to provide better visibility